egs/wsj/s5/utils/lang/limit_arpa_unk_history.py

```diff
@@ -0,0 +1,144 @@
+    
+    if len(ngram_counts) == 0:
+        sys.exit("Couldn't get counts per ngram section. The input doesn't seem to be a valid ARPA language model.")
+
+    max_ngrams = list(ngram_counts.keys())[-1]
+    skip_rows = ngram_counts[1] 
+
+    if max_ngrams > 5: 
+        sys.exit("This script supports up to 5-gram language models.")
+
+    return max_ngrams, skip_rows, ngram_counts
+
+
+def find_and_replace_unks(old_lm_lines,max_ngrams,skip_rows):
+    ngram_diffs =  defaultdict(int)
+    unk_pattern = re.compile("[0-9.-]+(?:[\s\\t]\S+){1,3}[\s\\t]"+args.oov_dict_entry+"[\s\\t](?!-[0-9]+\.[0-9]+).*")
+    backoff_pattern = re.compile("[0-9.-]+(?:[\s\\t]\S+){1,3}[\s\\t]<unk>[\s\\t]-[0-9]+\.[0-9]+")
+    passed_2grams, last_ngram = False, False
+    unk_row_count, backoff_row_count = 0, 0
+
+    sys.stderr.write("Upadting the language model .. \n")
+    new_lm_lines = old_lm_lines[:skip_rows]
+
+    for i in range(skip_rows, len(old_lm_lines)):
+            l = old_lm_lines[i].strip()
+
+            if "\{}-grams:".format(3) in l:
+                passed_2grams = True
+            if "\{}-grams:".format(max_ngrams) in l:
+                last_ngram = True
+                
+            # remove any n-gram states of the form: foo <unk> -> X 
+            # that is, any n-grams of order > 2 where <unk> is the second-to-last word
+            # here we skip 1-gram and 2-gram sections of arpa 
+
+            if passed_2grams:
+                g_unk = unk_pattern.search(l)
+                if g_unk:
+                    ngram = len(g_unk.group(0).split()) - 1 
+                    ngram_diffs[ngram] = ngram_diffs[ngram] - 1
+                    unk_row_count += 1
+                    continue 
+            
+            # remove backoff probability from the lines that end with <unk>
+            # for example, the -0.64 in -4.09 every <unk> -0.64
+            # here we skip the last n-gram section because it doesn't include backoff probabilities
+
+            if not last_ngram:  
+                g_backoff = backoff_pattern.search(l)
+                if g_backoff:
+                    updated_row = g_backoff.group(0).split()[:-1]
+                    updated_row = updated_row[0]+"\t"+(" ".join(updated_row[1:]))+"\n"
+                    new_lm_lines.append(updated_row)
+                    backoff_row_count += 1
+                    continue
+                
+            new_lm_lines.append(l+"\n")
+
+    sys.stderr.write("Removed {} lines including {} as second-to-last term\n".format(unk_row_count, args.oov_dict_entry))
+    sys.stderr.write("Removed backoff probabilties from {} lines\n".format(backoff_row_count))
+
+    return new_lm_lines, ngram_diffs
+
+
+def read_old_lm():
+    sys.stderr.write("Reading ARPA LM frome input stream .. \n")
+    
+    if args.compressed:
+        input_stream = gzip.open(sys.stdin.buffer, 'rt')
+    else:
+        input_stream = io.TextIOWrapper(sys.stdin.buffer) #encoding="latin-1"
```

Comment:
Your problem with the latin-1 may be that you are not doing the encoding to both input and output streams.
Please do it and if it doesn't work, let me know specifically what the error is.